### PR TITLE
Jvenstad/fix config model inconsitency

### PR DESCRIPTION
**Pull request by jonmv**
@hmusum please review.

This should fix the race we've seen where config super models get out of sync across config servers. 

The last commit is the most interesting one, where the synchronisation of the critical tasks is added. 
1. Creating, deleting or changing the active session of an application can only be done while its lock is held.
2. Reloading config for an application, in `TenantRequestHandler`, is done under the same application lock, and verifies the application still exists, and that the session about to be activated is the active one.
3. Removing an application from the config super model is also done under the same application lock, but only if the application is still supposed to be removed.

### CODE REVIEW

**Comment by jonmv** [comment]
Think I'll add one more.

**Comment by jonmv** [comment]
I assume the NR lockwas held  during redeployment as a previous attempt at avoiding these races. However, since that lock isn't used by other code for other things than changing the state of the _nodes_ owned by the application, it didn't have much positive effect;  the most prominent was rather that external deployments would be delayed, or, eventually, fail due to lock acquisition timeout.

**Comment by hmusum** [comment]
Yes, I think that's correct now. Earlier on we had bugs in NR, though, which holding the lock may have helped avoid.

File: configserver/src/main/java/com/yahoo/vespa/config/server/application/TenantApplications.java [hunk 1]
```diff
@@ -88,52 +91,55 @@ public static TenantApplications create(Curator curator, ReloadHandler reloadHan
...
-    public long getSessionIdForApplication(ApplicationId applicationId) {
-        String path = applicationPath(applicationId).getAbsolute();
-        try {
-            return Long.parseLong(Utf8.toString(curator.framework().getData().forPath(path)));
-        } catch (Exception e) {
-            throw new IllegalArgumentException(TenantRepository.logPre(applicationId) + "Unable to read the session id from '" + path + "'", e);
-        }
+    public long requireActiveSessionOf(ApplicationId applicationId) {
+        return activeSessionOf(applicationId)
+                .orElseThrow(() -> new IllegalArgumentException("Application '" + applicationId + "' has no active session."));
```
Comment (hmusum) [inline configserver/src/main/java/com/yahoo/vespa/config/server/application/TenantApplications.java]: Seems to me that there are users of this method that will not handle an exception being thrown, e.g. ApplicationRepository.lastDeployTime()

File: configserver/src/main/java/com/yahoo/vespa/config/server/session/SessionPreparer.java [hunk 2]
```diff
@@ -113,9 +113,8 @@ public ConfigChangeActions prepare(SessionContext context, DeployLogger logger,
             log.log(LogLevel.DEBUG, () -> "time used " + params.getTimeoutBudget().timesUsed() +
                     " : " + params.getApplicationId());
             return preparation.result();
-        } catch (OutOfCapacityException e) {
-            throw e;
-        } catch (IllegalArgumentException e) {
+        }
```
Comment (hmusum) [inline configserver/src/main/java/com/yahoo/vespa/config/server/session/SessionPreparer.java]: We want to return "Out of capacity" in these cases, not "Invalid application".

**Comment by hmusum** [comment]
Generally looks good, a few comments.  

I haven't thought about upgrade/downgrade scenarios and whether we handle all changes to e.g. zookeeper paths. Have you?

File: configserver/src/main/java/com/yahoo/vespa/config/server/application/TenantApplications.java (hunk 1 above)
Comment (jonmv) [reply configserver/src/main/java/com/yahoo/vespa/config/server/application/TenantApplications.java]: This threw before, as well, when the application didn't have an active session. I'll check if that may have been unfortunate.

File: configserver/src/main/java/com/yahoo/vespa/config/server/session/SessionPreparer.java (hunk 2 above)
Comment (jonmv) [reply configserver/src/main/java/com/yahoo/vespa/config/server/session/SessionPreparer.java]: As `OutOfCapacityException` isn't an `IllegalArgumentException`, this is what happens now as well.

**Comment by jonmv** [comment]
I believe so :)

File: configserver/src/main/java/com/yahoo/vespa/config/server/session/SessionPreparer.java (hunk 2 above)
Comment (hmusum) [reply configserver/src/main/java/com/yahoo/vespa/config/server/session/SessionPreparer.java]: Ah, of course. Good.

**Pull request merged**